storage/innobase/xtrabackup/src/ds_decrypt.c

```diff
@@ -0,0 +1,845 @@
+			msg("decrypt: write to destination failed.\n");
+			err = TRUE;
+		}
+
+		crypt_file->bytes_processed += thd->from_len;
+
+		pthread_mutex_unlock(&thd->data_mutex);
+		pthread_mutex_unlock(&thd->ctrl_mutex);
+
+		crypt_file->buf_len = 0;
+
+		if (err) {
+			return 1;
+		}
+	}
+
+	while (parse_result == XB_CRYPT_READ_CHUNK && len > 0) {
+		uint max_thread;
+
+		for (i = 0; i < nthreads; i++) {
+			thd = threads + i;
+
+			pthread_mutex_lock(&thd->ctrl_mutex);
+
+			parse_result = parse_xbcrypt_chunk(
+				thd, buf, len, &bytes_processed);
+
+			if (parse_result == XB_CRYPT_READ_ERROR) {
+				pthread_mutex_unlock(&thd->ctrl_mutex);
+				err = TRUE;
+				break;
+			}
+
+			thd->parse_result = parse_result;
+
+			if (parse_result != XB_CRYPT_READ_CHUNK) {
+				pthread_mutex_unlock(&thd->ctrl_mutex);
+				break;
+			}
+
+			pthread_mutex_lock(&thd->data_mutex);
+			thd->data_avail = TRUE;
+			pthread_cond_signal(&thd->data_cond);
+			pthread_mutex_unlock(&thd->data_mutex);
+
+			len -= bytes_processed;
+			buf += bytes_processed;
+		}
+
+		max_thread = (i < nthreads) ? i :  nthreads - 1;
+
+		/* Reap and write decrypted data */
+		for (i = 0; i <= max_thread; i++) {
+			thd = threads + i;
+
+			if (thd->parse_result != XB_CRYPT_READ_CHUNK) {
+				break;
+			}
+
+			pthread_mutex_lock(&thd->data_mutex);
+			while (thd->data_avail == TRUE) {
+				pthread_cond_wait(&thd->data_cond,
+						  &thd->data_mutex);
+			}
+
+			if (thd->failed) {
+				msg("decrypt: failed to decrypt chunk.\n");
+				err = TRUE;
+			}
+
+			xb_a(thd->to_len > 0);
+
+			if (!err && ds_write(crypt_file->dest_file, thd->to,
+				     thd->to_len)) {
+				msg("decrypt: write to destination failed.\n");
+				err = TRUE;
+			}
+
+			crypt_file->bytes_processed += thd->from_len;
+
+			pthread_mutex_unlock(&thd->data_mutex);
+			pthread_mutex_unlock(&thd->ctrl_mutex);
+		}
+
+		if (err) {
+			return 1;
+		}
+	}
+
+	if (parse_result == XB_CRYPT_READ_INCOMPLETE && len > 0) {
+		crypt_file->buf_len = len;
+		if (crypt_file->buf_size < len) {
+			crypt_file->buf = (uchar *) my_realloc(
+					crypt_file->buf,
+					crypt_file->buf_len,
+					MYF(MY_FAE | MY_ALLOW_ZERO_PTR));
+			crypt_file->buf_size = len;
+		}
+		memcpy(crypt_file->buf, buf, len);
+	}
+
+	return 0;
+}
+
+static
+int
+decrypt_close(ds_file_t *file)
+{
+	ds_decrypt_file_t	*crypt_file;
+	ds_file_t		*dest_file;
+	int			rc = 0;
+
+	crypt_file = (ds_decrypt_file_t *) file->ptr;
+	dest_file = crypt_file->dest_file;
+
+	if (ds_close(dest_file)) {
+		rc = 1;
+	}
+
+	my_free(crypt_file->buf);
+	my_free(file);
+
+	return rc;
+}
+
+static
+void
+decrypt_deinit(ds_ctxt_t *ctxt)
+{
+	ds_decrypt_ctxt_t 	*crypt_ctxt;
+
+	xb_ad(ctxt->pipe_ctxt != NULL);
+
+	crypt_ctxt = (ds_decrypt_ctxt_t *) ctxt->ptr;
+
+	destroy_worker_threads(crypt_ctxt->threads, crypt_ctxt->nthreads);
+
+	my_free(ctxt->root);
+	my_free(ctxt);
+}
+
+static
+crypt_thread_ctxt_t *
+create_worker_threads(uint n)
```

Comment:
For separating threadpool from datasink logged https://blueprints.launchpad.net/percona-xtrabackup/+spec/datasink-threadpool